src/runtime_src/core/common/api/xrt_module.cpp

```diff
@@ -223,7 +238,10 @@ struct patcher
         break;
       case symbol_type::shim_dma_base_addr_symbol_kind:
         // new_value is a bo address
-        patch57(bd_data_ptr, new_value + item.offset_to_base_bo_addr);
+        if (m_os_abi == Elf_Amd_npu3)
+          patch57_npu3(bd_data_ptr, new_value + item.offset_to_base_bo_addr);
+        else
+          patch57(bd_data_ptr, new_value + item.offset_to_base_bo_addr);
```

Comment:
This OK. But I think differentiating this by a new symbol type is cleaner than os_abi. Conceptly, one symobl type should map to one patching schema

Reply:
I think @HimanshuChoudhary-Xilinx is trying to avoid adding more patching schema because currently we only use 3bits  for encoding schema-type.

Reply:
> I think @HimanshuChoudhary-Xilinx is trying to avoid adding more patching schema because currently we only use 3bits for encoding schema-type.

not 4 bits (0-3)?

Reply:
its not a different patching... shimbd in aie2ps and npu3 have different configuration
in aie2ps, host address get patched on 1st,2nd, 8th word of BD while in npu3 its on 0th and 1st word of BD.
it the same 57bit patching

Reply:
> its not a different patching... shimbd in aie2ps and npu3 have different configuration in aie2ps, host address get patched on 1st,2nd, 8th word of BD while in npu3 its on 0th and 1st word of BD. its the same 57bit patching

I would argue shim_dma_48 is not a different patching either. It just patch to second and third of 8 BD WORDs.

OS_ABI is used to determine ELF's architecture. What ELF spec we should follow.

Reply:
do we have enough symbol type to hold all potential patching method?
if so, then we need something like,
patch57_aie2ps
patch57_aie4
if not, we may need combination of symbol and os_abi, like what Himanshu did.

Reply:
Yes. We do have enough symbol type to hold all potential patching method for the moment.

Reply:
A Open question, if os_abi is about ELF spec, then how we differentiate between a elf with aie2ps controlcode and elf with aie4 controlcode because an elf with aie2ps controlcode will not work on aie4 device or vice-versa.
XRT should have a check on this.

Reply:
> A Open question, if os_abi is about ELF spec, then how we differentiate between a elf with aie2ps controlcode and elf with aie4 controlcode because an elf with aie2ps controlcode will not work on aie4 device or vice-versa. XRT should have a check on this.

I am totally fine if we define a different os_abi between aie2ps and aie4 if we need to. I am just saying we should use a new patching schema at this place instead of using os_abi.